src/protocol/attribute.rs

```diff
@@ -0,0 +1,224 @@
+use super::{AttributionModel, RecordId};
+use crate::{
+    error::BoxError,
+    field::Field,
+    helpers::mesh::{Gateway, Mesh},
+    protocol::context::ProtocolContext,
+    secret_sharing::Replicated,
+};
+use futures::future::try_join_all;
+
+/// Obilivious attribution protocol
+///
+/// Obliviously attribute trigger events to corresponding source events for different breakdowns in a MPC.
+/// There are several common attribution models such as Last Touch, Even Credit and Exponential Decay.
+/// https://github.com/patcg-individual-drafts/ipa/blob/main/IPA-End-to-End.md#oblivious-last-touch-attribution
+#[allow(dead_code)]
+pub struct Attribute<'a, G> {
+    ctx: &'a ProtocolContext<'a, G, AttributionModel>,
+    step: AttributionModel,
+}
+
+impl<'a, G> Attribute<'a, G> {
+    const _IS_TRIGGER_BIT: usize = 2;
+    const _HELPER_BIT: usize = 3;
+    const _VALUE: usize = 5;
+
+    #[allow(dead_code)]
+    pub fn new(ctx: &'a ProtocolContext<'a, G, AttributionModel>, step: AttributionModel) -> Self {
+        Self { ctx, step }
+    }
+
+    /// Executes the protocol and produces an attribution result for the model specified by `step` in `Attribute::new`.
+    /// Attribution result will be used to calculate the aggregation in the next step.
+    #[allow(dead_code)]
+    pub async fn execute<M, F>(
+        &self,
+        input: &'a [Vec<Replicated<F>>],
+    ) -> Result<Vec<Replicated<F>>, BoxError>
+    where
+        M: Mesh,
+        G: Gateway<M, AttributionModel>,
+        F: Field,
+    {
+        match self.step {
+            AttributionModel::LastTouch => self.last_touch_attribution(input).await,
+        }
+    }
+
+    async fn last_touch_attribution<M, F>(
+        &self,
+        input: &'a [Vec<Replicated<F>>],
+    ) -> Result<Vec<Replicated<F>>, BoxError>
+    where
+        M: Mesh,
+        G: Gateway<M, AttributionModel>,
+        F: Field,
+    {
+        // TOOD: input length range check. Can we assume `assert!(num_rows < u32::MAX)`?
```

Comment:
I would much rather prefer if we turn this assumption into a type, i.e. use u32 here and below. In this case we don't need to keep being cautious about pieces of code that assume range fits into u32 but does not validate it. 

This unfortunately leads to major inconveniences working with vector indices and vector capacity as we have to use usizes for it. 

@martinthomson @benjaminsavage I wonder if we should create a wrapper around Vec and implement `Index<u32>/IndexMut<u32>` for it? We will enforce that this vector never grows over $2^{32}-1$ elements but working with it will be much more ergonomic and type safe than standard vec.

Reply:
Yes, I think that a wrapper would do a lot to help.

Reply:
Oh, I should say: the index type should probably be a typedef, like `type RecordIndex = u32` or something along those lines.